virtool/users/data.py

```diff
@@ -292,91 +362,265 @@ async def update(self, user_id: str, data: UpdateUserRequest):
-                self._mongo,
-                user_id,
-                document["administrator"],
-                document["groups"],
-                merge_group_permissions(groups),
+    @emits(Operation.UPDATE)
+    async def update(self, user_id: str, data: UpdateUserRequest):
+        """
+        Update a user.
+
+        Sessions and API keys are updated as well.
+
+        :param user_id: the ID of the user to update
+        :param data: the update data object
+        :return: the updated user
+        """
+        async with both_transactions(self._mongo, self._pg) as (
+            mongo_session,
+            pg_session,
+        ):
+            user = (
+                await pg_session.execute(
+                    select(SQLUser).where(SQLUser.legacy_id == user_id).limit(1)
+                )
+            ).scalar()
+
+            if user is None:
+                return await self.update_mongo_only(user_id, data)
+
+            document = await self._mongo.users.find_one(
+                {"_id": user_id}, ["administrator", "groups"], session=mongo_session
             )
 
+            if document is None:
+                raise ResourceNotFoundError("User does not exist")
+
+            data = data.dict(exclude_unset=True)
+
+            updates = {}
+
+            if "administrator" in data:
+                updates["administrator"] = data["administrator"]
+                user.administrator = data["administrator"]
+
+                role_assignment = AdministratorRoleAssignment(
+                    user_id, AdministratorRole.FULL
+                )
+
+                if data["administrator"]:
+                    await self._authorization_client.add(role_assignment)
+                else:
+                    await self._authorization_client.remove(role_assignment)
+
+            if "force_reset" in data:
+                updates.update(
+                    {"force_reset": data["force_reset"], "invalidate_sessions": True}
+                )
+                user.force_reset = data["force_reset"]
+                user.invalidate_sessions = True
+
+            if "password" in data:
+                hashedpass = virtool.users.utils.hash_password(data["password"])
+                timestamp = virtool.utils.timestamp()
+                updates.update(
+                    {
+                        "password": hashedpass,
+                        "last_password_change": timestamp,
+                        "invalidate_sessions": True,
+                    }
+                )
+                user.password = hashedpass
+                user.last_password_change = timestamp
+                user.invalidate_sessions = True
+
+            if "groups" in data:
+                try:
+                    updates.update(
+                        await compose_groups_update(self._pg, data["groups"])
+                    )
+
+                    await pg_session.execute(
+                        delete(user_group_associations)
```

Comment:
You should be able to manage the the groups associated via SQLAlchemy User model ORM. Clearing all the associations and rewriting them seems clunky

Reply:
Did you have a chance to try this via ORM?